src/chains/ethereum/ethereum/src/helpers/simulation-handler.ts

```diff
@@ -0,0 +1,484 @@
+      );
+      fromAccount.balance.isub(txCost);
+      await stateManager.putAccount(caller, fromAccount);
+
+      this.#runCallOpts = {
+        caller,
+        data: transaction.data && transaction.data.toBuffer(),
+        gasPrice: new BN(transaction.gasPrice.toBuffer()),
+        gasLimit: new BN(Quantity.toBuffer(gasLeft)),
+        to,
+        value:
+          transaction.value == null
+            ? new BN(0)
+            : new BN(transaction.value.toBuffer()),
+        block: transaction.block as any
+      };
+    } else {
+      this.#initializationError = ERROR.OUT_OF_GAS;
+    }
+  }
+
+  #setupStepEventEmits = () => {
+    if (this.#emitEvents) {
+      this.#vm.on("step", (event: InterpreterStep) => {
+        if (!this.#emitStepEvent) return;
+        const ganacheStepEvent = makeStepEvent(this.#transactionContext, event);
+        this.#blockchain.emit("ganache:vm:tx:step", ganacheStepEvent);
+      });
+    }
+  };
+
+  #emitBefore = () => {
+    if (this.#emitEvents) {
+      this.#blockchain.emit("ganache:vm:tx:before", {
+        context: this.#transactionContext
+      });
+    }
+  };
+
+  #emitAfter = () => {
+    if (this.#emitEvents) {
+      this.#blockchain.emit("ganache:vm:tx:after", {
+        context: this.#transactionContext
+      });
+    }
+  };
+
+  public toLightEJSAddress(address?: Address): EthereumJsAddress {
+    if (address) {
+      const buf = address.toBuffer();
+      return { buf, equals: (a: { buf: Buffer }) => buf.equals(a.buf) } as any;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Runs `vm.runCall` based off of the options set from running `initialize`.
+   * Throws if there is an `exceptionError` on the result.
+   * @returns The `EVMResult` from running the transaction.
+   */
+  public async runCall(): Promise<EVMResult> {
+    let callResult: EVMResult;
+    if (this.#initializationError) {
+      callResult = {
+        execResult: {
+          runState: { programCounter: 0 },
+          exceptionError: new VmError(this.#initializationError),
+          returnValue: BUFFER_EMPTY
+        }
+      } as any;
+    } else if (this.#runCallOpts) {
+      callResult = await this.#vm.runCall(this.#runCallOpts);
+      this.#emitAfter();
+    } else {
+      throw new Error("Unknown error when simulating transaction.");
+    }
+
+    if (callResult.execResult.exceptionError) {
+      throw new CallError(callResult);
+    } else {
+      return callResult;
+    }
+  }
+
+  /**
+   * Creates an access list for the simulated transaction.
+   * @param initialAccessList An initial access list to use for comparison
+   * against the generated access list.
+   * @returns
+   */
+  public async createAccessList(initialAccessList: AccessList) {
+    await this.#stateManager.checkpoint();
+    return await this.#createAccessList(initialAccessList);
+  }
+
+  /**
+   * Recursively runs the `vm.runCall` function until the touched addresses and
```

Comment:
Everything is written the way you describe, my question is about the base case. 

Is there a guarantee that we will always reach
``` javascript
JSON.stringify(previousAccessList) === JSON.stringify(accessList)
```
before we run out of memory and get a `Error: Max Call Stack...`? I understand `stateManager` comes from `VM`, which comes from `ethereumjs` and I did not check to see if this is how they do it but I understand it could be just the way it is. 

If no guarantee, would it be possible to write it as a loop? I don't see anything that looks leaky I just catch myself making nervous engineer noises when I get into JS recursion because the tail call optimization features aren't really great compared to some languages. 

Also, if you _really_ like recursion I believe you can use a trampoline thunk (but I have never done it in production just in play) to implement tail end elimination to the function call and achieve memory safety.

https://raganwald.com/2013/03/28/trampolines-in-javascript.html

(I could also be misreading the function and you are already accounting for memory)

Reply:
This is a better example, the previous link goes more in-depth 
https://raganwald.com/2013/03/29/high-level-trampolining.html

Reply:
This is an interesting question. 

Each time the `#createAccessList` function is run, the VM runs the transaction and "warms" accounts and storage keys as they are touched. Then we use those warmed accounts/keys to generate an access list. 

So, running a transaction will always generate an access list (though it may be empty) and the same contract code path (aka the ordered list of opcodes being executed) will generate the same access list. Thus, if the transaction is run twice in a row with everything being identical, `JSON.stringify(previousAccessList) === JSON.stringify(accessList)` will be `true`.

However, contracts can have different code paths based off of the gas used at some certain step, which would be affected by the presence of an access list. I haven't been able to come up with a contract function that touches a new set of accounts every time it is run, forever. My solidity knowledge is pretty limited, though. 

Still, we could manually create this failure case where we have a contract that keeps touching more accounts if the previous one was warmed. This would cause the `Error: Max Call Stack...` error you're concerned about:
```Solidity
function multiAccessList(address addr1, address addr2, ...) public view {
  uint256 startGas = gasleft();
  uint256 bal = addr1.balance;
  uint256 gasUsed = startGas - gasleft();
  require(bal > 0, "Balance is less than 0");
  if(gasUsed == 100) {
    startGas = gasleft();
    bal = addr2.balance;
    gasUsed = startGas - gasleft();
    require(bal > 0, "Balance is less than 0");
    if(gasUsed == 100) {
      ...
    }
  }
} 
``` 

But I have no idea how many different accounts you'd have to touch and how many times the `#createAccessList` function would have to be called for that to happen. I don't know how at risk this is of crashing ganache in practice, but I can be thinking of a fix for it if we think it's necessary.

Reply:
I get what was being changed with the comparison yesterday. I was totally wrong. Since the access list is just two arrays, iterating is going to be wayyyy faster. Thanks for your patience